### PR TITLE
Jenkins Pipeline updates for Canceling Jobs

### PR DESCRIPTION
**Pull request by TerrenceMcGuinness-NOAA**
# Description

Tuning updates for Jenkins Pipeline :
- Added short circuit for all parallel runs of cases on error of any
- Fixed canceling of all scheduled jobs on first case error
- Added feature to save error log files to Jenkins Archive facility on fail

# Type of change

- Bug fix (fixes all kill scheduler on fail )
- New feature (Archive logs)
- Maintenance (CI/CD)

# Change characteristics
- Is this a breaking change (a change in existing functionality)? NO
- Does this change require a documentation update? NO

# How has this been tested?
Tested on development Jenkins multi-branch Pipeline and forced error

### CODE REVIEW

**Comment by TerrenceMcGuinness-NOAA** [comment]
**C48mx500_3DVarAOWCDA fails on Hera with missing gdas file:**
```
Terry.McGuinness (hfe06) ~ $ grep gdas.t18z.adt_rads_all.2021032418.nc4 /scratch1/NCEPDEV/global/Terry.McGuinness/GFS_CI_ROOT_JENKINS/Jenkins/workspace/flow_EMC-Global-Pipeline_PR-2307/TESTDIR/RUNTESTS/COMROOT/C48mx500_3DVarAOWCDA_be789b15/logs/2021032418/gdasprepoceanobs.log
FileNotFoundError: [Errno 2] No such file or directory: 'gdas.t18z.adt_rads_all.2021032418.nc4'
OSError: unable to copy gdas.t18z.adt_rads_all.2021032418.nc4 to /scratch1/NCEPDEV/global/Terry.McGuinness/GFS_CI_ROOT_JENKINS/Jenkins/workspace/flow_EMC-Global-Pipeline_PR-2307/TESTDIR/RUNTESTS/COMROOT/C48mx500_3DVarAOWCDA_be789b15/gdas.20210324/18/obs/gdas.t18z.adt_rads_all.2021032418.nc4
```

File: Jenkinsfile [hunk 1]
```diff
@@ -76,7 +76,11 @@ pipeline {
                                     env.MACHINE_ID = machine // MACHINE_ID is used in the build scripts to determine the machine and is added to the shell environment
                                     if (fileExists("${HOMEgfs}/sorc/BUILT_semaphor")) { // if the system is already built, skip the build in the case of re-runs
                                         sh(script: "cat ${HOMEgfs}/sorc/BUILT_semaphor", returnStdout: true).trim() // TODO: and user configurable control to manage build semphore
-                                        ws(commonworkspace) { pullRequest.comment("Cloned PR already built (or build skipped) on ${machine} in directory ${HOMEgfs}") }
+                                        pullRequest.comment("Cloned PR already built (or build skipped) on ${machine} in directory ${HOMEgfs}<br>Still doing a checkout to get the latest changes")
+                                        sh(script: 'source workflow/gw_setup.sh;git pull;git submodule update --init --recursive', returnStatus: true)
```
Comment (WalterKolczynski-NOAA) [inline Jenkinsfile]:
```suggestion
                                        sh(script: 'source workflow/gw_setup.sh; git pull --recurse-submodules', returnStatus: true)
```

File: Jenkinsfile [hunk 2]
```diff
@@ -89,17 +93,19 @@ pipeline {
                                             sh(script: "echo ${HOMEgfs} > BUILT_semaphor", returnStatus: true)
                                         }
                                     }
-                                    if (pullRequest.labels.any { value -> value.matches("CI-${Machine}-Building") }) {
-                                        pullRequest.removeLabel("CI-${Machine}-Building")
-                                }
-                                    pullRequest.addLabel("CI-${Machine}-Running")
-                            }
+                                    if (env.CHANGE_ID && system == 'gfs') {
```
Comment (WalterKolczynski-NOAA) [inline Jenkinsfile:96]: Why is `system == 'gfs' here?

File: ci/scripts/run-check_ci.sh [hunk 3]
```diff
@@ -77,15 +78,16 @@
...
     {
      echo "Error logs:"
      echo "${error_logs}"
-    } >> "${RUNTESTS}/ci.log"
-    sed -i "s/\`\`\`//2g" "${RUNTESTS}/ci.log"
-    sacct --format=jobid,jobname%35,WorkDir%100,stat | grep "${pslot}" | grep "${pr}\/RUNTESTS" |  awk '{print $1}' | xargs scancel || true
+    } | tee -a  "${run_check_logfile}"
+    # rm -f "${RUNTESTS}/error.logs"
+    for log in ${error_logs}; do
+      echo "${log}" | sed 's/.*RUNTESTS/RUNTESTS/' >> "${RUNTESTS}/error.logs"
```
[the remark this answers is not in the document]
Comment (WalterKolczynski-NOAA) [reply ci/scripts/run-check_ci.sh]:
```suggestion
      echo "${log//*RUNTESTS/RUNTESTS}" >> "${RUNTESTS}/error.logs"
```

**Review by WalterKolczynski-NOAA — CHANGES_REQUESTED** [review]

File: Jenkinsfile (hunk 2 above)
Comment (TerrenceMcGuinness-NOAA) [reply Jenkinsfile:96]: To avoid race condition for updating labels when running parallel builds.  This need to happen only once.

**Review by TerrenceMcGuinness-NOAA — COMMENTED** [review]
reviewed comments and commit changes with git pull recursive command on no rebuild

**Review by WalterKolczynski-NOAA — APPROVED** [review]

**Comment by WalterKolczynski-NOAA** [comment]
CI is failing for unrelated reasons, so we are going to merge this (which has no impact outside of CI) and address the problem separately.

**Pull request merged**